### ISSUE TITLE
export_hf: experimental_use_fp16 produces fp32 signature I/O

### ISSUE THREAD

**Issue by vieenrose** [google-ai-edge/litert-torch#1137]
**TL;DR:** In `export_hf`, setting `experimental_use_fp16=True` has no effect on the exported model's signature dtypes — every KV-cache and activation input/output still comes out `float32`, in both `model.tflite` and `model_quantized.tflite`.

## Repro

- Repo state: current `main` (checkout of 2026-07-28), transformers 5.14.1, Linux x86-64.
- Export: `litert_torch/generative/export_hf/export.py::export` with
  `model=google/gemma-4-E2B-it, split_cache=True, cache_length=16384, prefill_lengths=[128], experimental_use_fp16=True` (quantization left at the default `dynamic_wi8_afp32`).
- Export completes successfully.

## Expected

`export_hf/core/cache.py` builds the example KV tensors as fp16 when the flag is set:

```python
cache_dtype = (
    torch.float16 if export_config.experimental_use_fp16 else torch.float32
)
```

and `exportable_module.py` halves `inputs_embeds`, so the exported prefill/decode signatures should carry fp16 KV I/O (halving KV-cache RAM — 576 MiB → 288 MiB at cache_length=16384 for this model).

## Observed

Auditing the produced `.tflite` with the interpreter's signature API: all `kv_cache_k_*` / `kv_cache_v_*` / `kv_slice_*` inputs and outputs, and the activation I/O, are `float32` in both the fp and the quantized model. The flag's only surviving effect appears to be the `prefer_activation_type='fp32_fp16'` hint in `litert_lm_builder.py` (which only matters when bundling a `.litertlm`).

So the fp16 dtype chosen at trace time is lost somewhere in the torch.export → converter pipeline, and consumers of the raw `.tflite` (custom engines driving the signatures directly) cannot get fp16 KV I/O at all.

Happy to provide the full signature dump or test a patched branch.

**Comment by shivvaam0001** [comment]
Hi @vieenrose,

Thanks for flagging this issue.

I tried to reproduce the issue and observed that the split cache export code path indeed ignores `experimental_use_fp16`. It unconditionally uses float32 when allocating key/value cache tensors and does not cast inputs like `embeddings` and `pos_emb` to `float16`. Even when `experimental_use_fp16=True` is set, all exported signature inputs and outputs still come out as float32:

```
Number of subgraphs: 6

Signature: prefill_16
Inputs:
  embeddings: dtype=float32, shape=[ 1 16 32], subgraph=0
  kv_cache_k_0: dtype=float32, shape=[  1   1 128  16], subgraph=0
  kv_cache_v_0: dtype=float32, shape=[  1   1  16 128], subgraph=0
  mask_global: dtype=float32, shape=[  1   1  16 144], subgraph=0
  pos_emb_cos: dtype=float32, shape=[ 1 16  1 16], subgraph=0
  pos_emb_sin: dtype=float32, shape=[ 1 16  1 16], subgraph=0
Outputs:
  kv_slice_k_0: dtype=float32, shape=[ 1  1 16 16], subgraph=0
  kv_slice_v_0: dtype=float32, shape=[ 1  1 16 16], subgraph=0

Signature: decode
Inputs:
  embeddings: dtype=float32, shape=[ 1 16 32], subgraph=0
  kv_cache_k_0: dtype=float32, shape=[  1   1 128  16], subgraph=0
  kv_cache_v_0: dtype=float32, shape=[  1   1  16 128], subgraph=0
  mask_global: dtype=float32, shape=[  1   1  16 144], subgraph=0
  pos_emb_cos: dtype=float32, shape=[ 1 16  1 16], subgraph=0
  pos_emb_sin: dtype=float32, shape=[ 1 16  1 16], subgraph=0
Outputs:
  kv_slice_k_0: dtype=float32, shape=[ 1  1  1 16], subgraph=1
  kv_slice_v_0: dtype=float32, shape=[ 1  1 16  1], subgraph=1
  logits: dtype=float32, shape=[   1    1 1000], subgraph=1

```

If there are any additional details you can share, that would be very helpful to us.
Thanks!

**Comment by tf-marissaw** [comment]
From looking at the code, I agree that `experimental_use_fp16` was intended to trigger `prefer_activation_type=fp32_fp16`. Perhaps a better name for this flag would have been `experimental_use_mixed_precision`.

The good news is that models will run with fp16 precision on the GPU by default in LiteRT and LiteRT-LM. Also, the KV cache is typically stored with fp16 precision by default unless the model has a quantized KV cache which doesn't appear to be the case here. Typically this conversion from fp32 in the .tflite or .litertlm file to fp16 happens _at runtime_ for the GPU. 

The reason `fp32_fp16` was introduced was because fp16 is not appropriate for all ops. Some operations like RoPE require fp32 to retain good precision. A naive conversion of all ops to fp16 at runtime was resulting in poor accuracy so `fp32_fp16` was added to allow models to mostly run in fp16.

If you would like to try running the model with fp32 to see what the difference in performance and memory usage is, you could export your model without `experimental_use_fp16`  and then set the [`force_f32`](https://github.com/google-ai-edge/LiteRT-LM/blob/c95b30a915275963a045dfe2948d8d59d3e849f5/runtime/engine/shared_flags.cc#L72) flag in LiteRT-LM at runtime.

**Comment by sirakiin** [comment]
Yes experimental_use_fp16 is meant for mixed precision activation, not for weights. Also right now it only works with externalize_embedder=True and single_token_embedder=True, and GPU only, after setting those flags you should see kv cache be fp16 dtype.

**Comment by aravindmurali22** [comment]
Hi @vieenrose , Closing this issue as the behavior and required flags (externalize_embedder=True and single_token_embedder=True) were clarified above. Feel free to reopen if you have further questions!
